### DISCUSSION TITLE
Webhooks no work: error 400

### DISCUSSION THREAD

**Discussion by DoAgency** [impress-org/givewp#6842] [Developer Q&A]
We installed "webhooks plugin" but we are encountering some difficulties.
After 2-3 call, the plugin stop working.
Plus, when it is functioning, the calls that it does goes in 400 error.

In function.php we put this code to test WPGive webhooks:

```
function prefix_get_private_data_permissions_check() {
    return true;
}
add_action('rest_api_init', function () {
        register_rest_route( 'api/v1', '/citta', array(
            'methods' => 'POST',
			'content-type' => 'application/json; charset=UTF-8',
            'callback' => 'create_city_from_data',
			'permission_callback' => 'prefix_get_private_data_permissions_check'	
        ));
});
function create_city_from_data($req) {
		error_log("ciao");
        $response = print_r($req, true);

        $res = new WP_REST_Response($response);
        $res->set_status(200);
		error_log(print_r($res, true));
        return ['req' => $res];
}
```

If I call my api by Postman everything works very well: https://[my-domain]/wp-json/api/v1/citta
If I call my code by webhooks of WPGive, for example I set "donation_updated", in log of  WPGive we see this error:
`{\"code\":\"rest_invalid_json\",\"message\":\"È stato passato un JSON con il body non valido.\",\"data\":{\"status\":400,\"json_error_code\":4,\"json_error_message\":\"Syntax error\"}}`

We cannot modify the body so we don't know how to proceed.

Anyway, after 2 times, also if I update a donation, there is no more log in WPGive.

**Comment by glaubersilva** [comment]
Hi @DoAgency 

We released the Webhooks addon version `1.0.2` yesterday which implements a fix to send event data in JSON format. Can you update to this last version and check if the problem has been solved?

**Comment by DoAgency** [comment]
Hi @glaubersilva,
thanks for your answer.
Unfortunately I have updated addon, but there is the same problem.
When there is the event, there is no log in log area in backoffice WpGive in Wordpress, and there is no log in wp-content/debug.log, not even about my error_log that you can see in my code.
It is like if event triggers nothing.

Let me know please.
Thanks
Regards

**Comment by glaubersilva** [comment]
@DoAgency Is possible that you have some problems in the `wp-cron` as the Addon uses `Action Scheduler`, which relies on the `wp-cron`, to schedule the requests. 

If that is the case, you can check the pending requests in the `WP > Tools > Scheduled Actions` screen. There you can trigger the pending requests manually. 

For reference, check the screenshot:

![pending-requests](https://github.com/impress-org/givewp/assets/16308864/0d0b1d82-10da-4250-8190-56ba2a01e481)

You also can install the [WP Crontrol plugin](https://wordpress.org/plugins/wp-crontrol/) to check possible problems with your `wp-cron` and ask your host to fix that if this is the case.

About the logs, keep in mind that our log system will record only errors by default. To log any other kind of messages like success messages, you need to set the `WP_DEBUG_LOG` const on your wp-config file, something like this setup:

```
// Enable WP_DEBUG mode
define( 'WP_DEBUG', true );

// Enable Debug logging to the /wp-content/debug.log file
define( 'WP_DEBUG_LOG', true );

// Disable display of errors and warnings
define( 'WP_DEBUG_DISPLAY', false );
@ini_set( 'display_errors', 0 );

// Use dev versions of core JS and CSS files (only needed if you are modifying these core files)
define( 'SCRIPT_DEBUG', false );
```

Reference: https://wordpress.org/documentation/article/debugging-in-wordpress/

Let me know if it did help figure out what is happening on your side. :)

**Comment by DoAgency** [comment]
Hi @glaubersilva and thanks for answer,

1) sorry but in my "Scheduled Actions" I have no a hook with name "givewp_webhooks_requests". Like you can see here:
![Senza titolo-1](https://github.com/impress-org/givewp/assets/38939209/d79889d4-754d-4cb2-b977-11399891ad45)

2) You say "our log system will record only errors by default" , but I see not just errors, I see also other information:
![Senza titolo-2](https://github.com/impress-org/givewp/assets/38939209/e9b06932-6a6b-460d-a7c1-7b0cda94acf1)

3) Here you can see the error 400 in log of  WPGive:
![Senza titolo-4](https://github.com/impress-org/givewp/assets/38939209/ab911d1b-8a1c-4926-be7b-2ebf10bd9a71)

4) Yes I know about WP_DEBUG_LOG, but no print my error_log that you can see in my code. It is like WpGive event trigers nothing.

Thanks
Regards

**Comment by glaubersilva** [comment]
@DoAgency thanks for sharing more details.

If no requests are created in the `WP > Tools > Scheduled Actions` screen, means that the events are not being triggered and it can happen due to a few reasons:

1. Make sure your are using the last GiveWP version;
2. The endpoint may be deactivated, please double-check on `Give > Tools > Webhooks` screen;
3. The endpoint may be is not be set up to listen to the triggered event, please double-check on `Give > Tools > Webhooks` screen;
4. You are trying to test an update event but are not updating any information that triggers the event.

For this last case, I realize that you mentioned in one of your replies that the `donation_updated` event is not trigger anything. However, this event will be triggered only if you update any information watched by this event. 

For this event in specific, these are the properties observed by the event: 

```
'firstName',
'lastName',
'email',
'amount',
'feeAmountRecovered',
'type',
```

Only these properties are watched because just they are sent in the event data body, so if you change any other information that is not sent by the event, it will not be triggered.

I also ran some tests with your code and it worked as expected. To test your code, I created a WP Rest API endpoint in a live site using your code and I set up my local test site to send events to this live site. 

The result you can see in the attached screenshots:

![donation_deleted](https://github.com/impress-org/givewp/assets/16308864/0542919d-250d-4bb8-a59b-e5f1c8789f18)
![donation_updated](https://github.com/impress-org/givewp/assets/16308864/0b2150dd-ba68-4a2f-8d61-bf6afe0969d1)

Let me know your findings!

**Comment by DoAgency** [comment]
Hi @glaubersilva ,
thanks for answer.

1) Like you can see here, we have last version of plugins:
![Senza titolo-1](https://github.com/impress-org/givewp/assets/38939209/c95a3508-22da-4701-9851-8bfbf6515d4d)

2) End point is active:
![Senza titolo-2](https://github.com/impress-org/givewp/assets/38939209/06bbb00e-2e56-453c-8554-1cef07bf299d)

3) Events are three. 
![Senza titolo-3](https://github.com/impress-org/givewp/assets/38939209/f1dc3264-e296-4619-9561-a400e3f3bae7)
For example is I use "donation_created", so I go to create a new donation like this:
![Senza titolo-5](https://github.com/impress-org/givewp/assets/38939209/bf6952b5-55fa-46a2-866d-5c68cd34ebc3)

But i have: 
- no log in WPGive area on backoffice.
- no log in wp-content/debug.log.
- no "givewp_webhooks_requests" in scheduled action.

**Comment by glaubersilva** [comment]
@DoAgency Thanks for sharing more details. 

Seems that you are using the manual donations addon to create new donations, so I ran a few tests and could confirm that the `donation_created` event is not being triggered when the donation is manually created. 

However, using a form to create the donations, it get triggered. Can you replicate this behavior on your end?

**Comment by DoAgency** [comment]
Hello @glaubersilva ,
now it works. yes the problem was that if i create donation by manual is not triggered the event.
Thanks

**Comment by matheuswd** [comment]
Hey, @DoAgency.

​We've investigated it and found that you've uncovered an edge-case issue with our Webhokos add-on. What is happening is that we are not triggering the donation_created event when a donation is created manually using the Manual Donations add-on.

I've escalated this to our development team so they can work on a fix. You can follow up on the progress of the fix here https://feedback.givewp.com/bug-reports/p/the-webhooks-add-on-should-trigger-the-donationcreated-event-when-using-manual-d. 

​​We are actively tracking user issues to make sure our dev team is aware of them and can get them fixed. You can use the link to check in for updates or to provide additional details of your own.

Give the post above a look, and let me know if you would like me to add any additional details. I want to make sure that you're able to see your needs as accurately reflected as possible, and I'm happy to include an internal note to our teams with any specific information you would like them to have.

The workaround is to create these donations directly on the donation form. Bear in mind that you must be logged out to do so; otherwise, you will mix donor profiles.
 
Thanks for using GiveWP! I hope you have a great day.
